pkg/crawl/report.go

```diff
@@ -122,24 +226,37 @@ func (r *Reporter) Collect(ctx context.Context, path string, d fs.DirEntry, resu
 	}
 }
 
-func (r *Reporter) buildCVEMessage(versions []string) string {
-	if r.DisableCVE45105 {
-		return "CVE-2021-45046 detected"
+func (r *Reporter) matchedCVEs(versions []string) []string {
+	if len(versions) == 0 {
+		return []string{"unknown version - unknown CVE status"}
 	}
-	if cve45105VersionsOnly(versions) {
-		return "CVE-2021-45105 detected"
-	}
-	return "CVE-2021-45046 and CVE-2021-45105 detected"
-}
-
-func cve45105VersionsOnly(versions []string) bool {
-	if len(versions) == 1 && (versions[0] == "2.16.0" || versions[0] == "2.12.2") {
-		return true
+	cvesFound := make(map[string]struct{})
+	for _, version := range versions {
+		major, minor, patch, parsed := ParseLog4jVersion(version)
+		if !parsed {
+			cvesFound["unknown version - unknown CVE status"] = struct{}{}
+		}
+		for _, vulnerability := range cveVersions {
+			if major >= vulnerability.FixedAfter.Major && minor >= vulnerability.FixedAfter.Minor && patch >= vulnerability.FixedAfter.Patch {
+				continue
+			}
+			vulnerable := true
+			for _, fixedVersion := range vulnerability.PatchedVersions {
+				if major == fixedVersion.Major && minor == fixedVersion.Minor && patch >= fixedVersion.Patch {
+					vulnerable = false
+					break
+				}
+			}
+			if vulnerable && !(r.DisableCVE45105 && vulnerability.CVE == "CVE-2021-45105") && !(r.DisableCVE44832 && vulnerability.CVE == "CVE-2021-44832") {
+				cvesFound[vulnerability.CVE] = struct{}{}
+			}
+		}
 	}
-	if len(versions) == 2 && versions[0] == "2.12.2" && versions[1] == "2.16.0" {
-		return true
+	var uniqueCVEs []string
+	for cve := range cvesFound {
+		uniqueCVEs = append(uniqueCVEs, cve)
 	}
-	return false
+	return uniqueCVEs
```

Comment:
nit: Add a `sort.Strings(uniqueCVEs)` before returning so the result is deterministic